### PR TITLE
Migrate Species Selector components to redux hooks

### PR DESCRIPTION
**Pull request by azangru**
## Related JIRA Issue(s)
https://www.ebi.ac.uk/panda/jira/browse/ENSWBSITES-1357

## Description
Code refactoring: convert components in Species Selector to use redux hooks.

## Deployment URL
http://refactor-species-selector-to-hooks.review.ensembl.org

### CODE REVIEW

File: src/ensembl/src/content/app/species-selector/components/species-commit-button/SpeciesCommitButton.test.tsx [hunk 1]
```diff
@@ -30,11 +35,26 @@ jest.mock(
...
+
+const defaultReduxState = {
+  speciesSelector: {
+    currentItem: committedHuman
+  }
+};
+
+const mockStore = configureMockStore([thunk]);
+
+const renderComponent = (state?: unknown) => {
```
Comment (jyothishnt) [inline src/ensembl/src/content/app/species-selector/components/species-commit-button/SpeciesCommitButton.test.tsx:51]: I see state typed as any in PopularSpeciesButton.test where as here it is unknown. I dont know which is better.

File: src/ensembl/src/content/app/species-selector/components/species-selector-app-bar/SpeciesSelectorAppBar.test.tsx [hunk 2]
```diff
@@ -42,46 +45,61 @@ jest.mock(
...
+const defaultReduxState = {
+  speciesSelector: {
+    committedItems: selectedSpecies
+  }
+};
+
+const mockStore = configureMockStore([thunk]);
+
+const renderComponent = (state?: any) => {
+  state = Object.assign({}, defaultReduxState, state);
```
Comment (jyothishnt) [inline src/ensembl/src/content/app/species-selector/components/species-selector-app-bar/SpeciesSelectorAppBar.test.tsx:59]: `merge` or `Object.assign` better?  Both gives same result, so is it just a personal coding preference?

File: src/ensembl/src/content/app/species-selector/components/species-selector-app-bar/SpeciesSelectorAppBar.tsx [hunk 3]
```diff
@@ -44,10 +39,16 @@ const PlaceholderMessage = () => (
...
+  const selectedSpecies = useSelector(getCommittedSpecies);
+  const dispatch = useDispatch();
+
   const mainContent =
-    props.selectedSpecies.length > 0 ? (
-      <SelectedSpeciesList {...props} />
+    selectedSpecies.length > 0 ? (
+      <SelectedSpeciesList
+        selectedSpecies={selectedSpecies}
+        dispatch={dispatch}
```
Comment (jyothishnt) [inline src/ensembl/src/content/app/species-selector/components/species-selector-app-bar/SpeciesSelectorAppBar.tsx:50]: trying to understand the code, what is it dispatching here?

File: src/ensembl/src/content/app/species-selector/components/species-selector-app-bar/SpeciesSelectorAppBar.test.tsx (hunk 2 above)
Comment (azangru) [reply src/ensembl/src/content/app/species-selector/components/species-selector-app-bar/SpeciesSelectorAppBar.test.tsx:59]: > Both gives same result

They don't. Lodash `merge` does a deep merge. Object.assign does a shallow merge.

Example: given

```
const a = {
  foo: 1,
  bar: { x: 1 },
  baz: [ 1, 2 ]
};

const b = {
  foo: 2,
  bar: { y: 2 },
  baz: []
}
```

then `merge({}, a, b)` will produce:

```
{
  foo: 2,
  bar: { x: 1, y: 2 },
  baz: [ 1, 2 ]
}
```

while `Object.assign({}, a, b)` will produce:

```
{
  foo: 2,
  bar: { y: 2 },
  baz: []
}
```

P.S.: both lodash's `merge` and Object.assign mutate the first object that they receive. `merge` from `lodash/fp` will produce a new object; but we are rarely using `lodash/fp` (mostly just for the `set` function).

File: src/ensembl/src/content/app/species-selector/components/species-commit-button/SpeciesCommitButton.test.tsx (hunk 1 above)
Comment (azangru) [reply src/ensembl/src/content/app/species-selector/components/species-commit-button/SpeciesCommitButton.test.tsx]: Generally, `unknown` is better. In the case of these tests though, both are equally bad 😄 

I was lazy and didn't want to define those types. You made me ashamed and I updated the type definitions with proper ones 😄

File: src/ensembl/src/content/app/species-selector/components/species-commit-button/SpeciesCommitButton.test.tsx [hunk 4]
```diff
@@ -15,12 +15,17 @@
...
+import thunk from 'redux-thunk';
 import { render } from '@testing-library/react';
 import userEvent from '@testing-library/user-event';
+import configureMockStore from 'redux-mock-store';
+import merge from 'lodash/merge';
+
+import * as speciesSelectorActions from 'src/content/app/species-selector/state/speciesSelectorActions';
 
 import { SpeciesCommitButton } from './SpeciesCommitButton';
 import { createSelectedSpecies } from 'ensemblRoot/tests/fixtures/selected-species';
```
[the remark this answers is not in the document]
Comment (azangru) [reply src/ensembl/src/content/app/species-selector/components/species-commit-button/SpeciesCommitButton.test.tsx]: ✅

File: src/ensembl/src/content/app/species-selector/components/popular-species-button/PopularSpeciesButton.test.tsx [hunk 5]
```diff
@@ -67,45 +107,60 @@ describe('<PopularSpeciesButton />', () => {
...
-      const { container } = render(<PopularSpeciesButton {...commonProps} />);
+    it('selects the species received from props when clicked', () => {
+      jest.spyOn(speciesSelectorActions, 'handleSelectedSpecies');
+
+      const { container } = renderComponent({ props: defaultProps });
       const button = container.querySelector(
         '.popularSpeciesButton'
       ) as HTMLElement;
-      const speciesData = commonProps.species;
+      const speciesData = defaultProps.species;
```
Comment (imransl) [inline src/ensembl/src/content/app/species-selector/components/popular-species-button/PopularSpeciesButton.test.tsx:143]: Do we need this variable?

File: src/ensembl/src/content/app/species-selector/components/species-selector-app-bar/SpeciesSelectorAppBar.tsx [hunk 6]
```diff
@@ -61,13 +62,16 @@ export const SpeciesSelectorAppBar = (props: Props) => {
   );
 };
 
-const SelectedSpeciesList = (props: Props) => {
+const SelectedSpeciesList = (props: {
+  selectedSpecies: CommittedItem[];
+  dispatch: Dispatch<any>;
```
[the remark this answers is not in the document]
Comment (azangru) [reply src/ensembl/src/content/app/species-selector/components/species-selector-app-bar/SpeciesSelectorAppBar.tsx]: I have moved useDispatch inside the `SelectedSpeciesList` instead 😄

**Review by jyothishnt — APPROVED** [review]

**Pull request merged**